File: typescript/tests/sonars.test.ts
/**
 * Unit tests for the sonars namespace — no live API required.
 *
 * Verifies the SDK matches the fldaily /sonars contract: method, path, query
 * params (undefined dropped), JSON bodies (incl. the explicit-null daily_cap on
 * PATCH), the `before` cursor walk, and 204 -> {} on delete. Uses an injected mock `fetch`.
 */
import { test } from 'node:test';
import assert from 'node:assert';

import { SurfClient } from '../src/index';

const BASE = 'https://api.surf.social/v1';
const SPEC = { subject: { hashtags: ['#opensearch'] }, surfaces: ['bluesky', 'mastodon'] };

interface Call { url: string; method?: string; body?: string }

function clientWithCapture(calls: Call[], responses: Array<{ status?: number; json?: unknown }> = []) {
  let i = 0;
  const mockFetch = (async (url: any, opts: any) => {
    calls.push({ url: String(url), method: opts?.method, body: opts?.body });
    const r = responses[i++] ?? {};
    const status = r.status ?? 200;
    return {
      ok: status >= 200 && status < 300,
      status,
      statusText: 'OK',
      headers: { get: () => null },
      json: async () => r.json ?? {},
      body: { cancel: async () => {} },
    } as any;
  }) as unknown as typeof fetch;
  return new SurfClient({ apiKey: 'surf_sk_live_k', fetch: mockFetch });
}

test('create POSTs name, spec and only the delivery fields given', async () => {
  const calls: Call[] = [];
  const c = clientWithCapture(calls);
  await c.sonars.create({ name: 'OpenSearch chatter', spec: SPEC });
  assert.equal(calls[0].method, 'POST');
  assert.equal(calls[0].url, `${BASE}/sonars`);
  assert.deepEqual(JSON.parse(calls[0].body as string), { name: 'OpenSearch chatter', spec: SPEC });
  await c.sonars.create({ name: 'n', spec: SPEC, enabled: false, cadence: 'instant', channels: [{ type: 'push' }], daily_cap: 20 });
  assert.deepEqual(JSON.parse(calls[1].body as string), {
    name: 'n', spec: SPEC, enabled: false, cadence: 'instant', channels: [{ type: 'push' }], daily_cap: 20,
  });
});

test('list / get / delete paths; ids are escaped as one segment; delete tolerates 204', async () => {
  const calls: Call[] = [];
  const c = clientWithCapture(calls, [{}, {}, {}, { status: 204 }]);
  await c.sonars.list();
  assert.equal(calls[0].method, 'GET');
  assert.equal(calls[0].url, `${BASE}/sonars`);
  await c.sonars.get('01j0ksyw6tgwfs0zmhhbc42kwa');
  assert.equal(calls[1].url, `${BASE}/sonars/01j0ksyw6tgwfs0zmhhbc42kwa`);
  await c.sonars.get('odd/id');
  assert.equal(calls[2].url, `${BASE}/sonars/odd%2Fid`);
  const out = await c.sonars.delete('abc');
  assert.equal(calls[3].method, 'DELETE');
  assert.equal(calls[3].url, `${BASE}/sonars/abc`);
  assert.deepEqual(out, {});
});

test('update PATCHes only the fields given and can clear daily_cap with an explicit null', async () => {
  const calls: Call[] = [];
  const c = clientWithCapture(calls);
  await c.sonars.update('abc', { name: 'renamed' });
  assert.equal(calls[0].method, 'PATCH');
  assert.equal(calls[0].url, `${BASE}/sonars/abc`);
  assert.deepEqual(JSON.parse(calls[0].body as string), { name: 'renamed' });
  await c.sonars.update('abc', { daily_cap: null, enabled: false });
  assert.deepEqual(JSON.parse(calls[1].body as string), { daily_cap: null, enabled: false });
});

test('matches drops absent params and passes before/limit', async () => {
  const calls: Call[] = [];
  const c = clientWithCapture(calls);
  await c.sonars.matches('abc');
  assert.equal(calls[0].url, `${BASE}/sonars/abc/matches`);
  await c.sonars.matches('abc', { before: 41, limit: 10 });
  assert.equal(calls[1].url, `${BASE}/sonars/abc/matches?before=41&limit=10`);
});

test('iterMatches follows next_before until it is null and honours a limit', async () => {
  const calls: Call[] = [];
  const c = clientWithCapture(calls, [
    { json: { matches: [{ id: 3, post_id: 'p3' }, { id: 2, post_id: 'p2' }], next_before: 2 } },
    { json: { matches: [{ id: 1, post_id: 'p1' }], next_before: null } },
  ]);
  const got: string[] = [];
  for await (const m of c.sonars.iterMatches('abc')) got.push(m.post_id);
  assert.deepEqual(got, ['p3', 'p2', 'p1']);
  assert.equal(calls.length, 2);
  assert.equal(calls[1].url, `${BASE}/sonars/abc/matches?before=2`);

  const calls2: Call[] = [];
  const c2 = clientWithCapture(calls2, [{ json: { matches: [{ id: 3, post_id: 'p3' }, { id: 2, post_id: 'p2' }], next_before: 2 } }]);
  const first: string[] = [];
  for await (const m of c2.sonars.iterMatches('abc', { limit: 1 })) first.push(m.post_id);
  assert.deepEqual(first, ['p3']);
  assert.equal(calls2[0].url, `${BASE}/sonars/abc/matches?limit=1`);
});

test('preview POSTs the spec with days as a query param', async () => {
  const calls: Call[] = [];
  const c = clientWithCapture(calls);
  await c.sonars.preview(SPEC);
  assert.equal(calls[0].method, 'POST');
  assert.equal(calls[0].url, `${BASE}/sonars/preview`);
  assert.deepEqual(JSON.parse(calls[0].body as string), SPEC);
  await c.sonars.preview(SPEC, { days: 7 });
  assert.equal(calls[1].url, `${BASE}/sonars/preview?days=7`);
});

File: go/sonars_test.go
package surf

import (
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"testing"
)

// Unit tests for the Sonars namespace — no live API required. An httptest
// server stands in for api.surf.social and records what the SDK sends: method,
// path (raw, so escaping is visible), query and body.

type sonarCall struct {
	method, rawPath, query string
	body                   []byte
}

func newSonarServer(t *testing.T, responses ...string) (*Client, *[]sonarCall) {
	t.Helper()
	calls := &[]sonarCall{}
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		body, _ := io.ReadAll(r.Body)
		*calls = append(*calls, sonarCall{r.Method, r.URL.EscapedPath(), r.URL.RawQuery, body})
		w.Header().Set("Content-Type", "application/json")
		if r.Method == "DELETE" {
			w.WriteHeader(http.StatusNoContent)
			return
		}
		resp := "{}"
		if n := len(*calls) - 1; n < len(responses) {
			resp = responses[n]
		}
		_, _ = w.Write([]byte(resp))
	}))
	t.Cleanup(srv.Close)
	c := NewClient("surf_sk_live_k")
	c.BaseURL = srv.URL
	return c, calls
}

var sonarSpec = SonarSpec{Subject: SonarSubject{Hashtags: []string{"#opensearch"}}, Surfaces: []string{"bluesky", "mastodon"}}

func TestSonarsCreateSendsNameSpecAndOnlyGivenDeliveryFields(t *testing.T) {
	c, calls := newSonarServer(t)
	if _, err := c.Sonars.Create(SonarRequest{Name: "OpenSearch chatter", Spec: &sonarSpec}); err != nil {
		t.Fatal(err)
	}
	got := (*calls)[0]
	if got.method != "POST" || got.rawPath != "/v1/sonars" {
		t.Errorf("got %s %s, want POST /v1/sonars", got.method, got.rawPath)
	}
	var body map[string]interface{}
	_ = json.Unmarshal(got.body, &body)
	if body["name"] != "OpenSearch chatter" || body["spec"] == nil {
		t.Errorf("body = %s", got.body)
	}
	for _, k := range []string{"enabled", "cadence", "channels", "daily_cap"} {
		if _, present := body[k]; present {
			t.Errorf("%s should be omitted when unset: %s", k, got.body)
		}
	}
	cap := 20
	enabled := false
	if _, err := c.Sonars.Create(SonarRequest{Name: "n", Spec: &sonarSpec, Enabled: &enabled, Cadence: "instant",
		Channels: []SonarChannel{{Type: "push"}}, DailyCap: &cap}); err != nil {
		t.Fatal(err)
	}
	if string((*calls)[1].body) != `{"name":"n","spec":{"subject":{"hashtags":["#opensearch"]},"surfaces":["bluesky","mastodon"]},"enabled":false,"cadence":"instant","channels":[{"type":"push"}],"daily_cap":20}` {
		t.Errorf("full body = %s", (*calls)[1].body)
	}
}

func TestSonarsListGetDeletePathsAndEscaping(t *testing.T) {
	c, calls := newSonarServer(t)
	if _, err := c.Sonars.List(); err != nil {
		t.Fatal(err)
	}
	if _, err := c.Sonars.Get("01j0ksyw6tgwfs0zmhhbc42kwa"); err != nil {
		t.Fatal(err)
	}
	if _, err := c.Sonars.Get("odd/id"); err != nil {
		t.Fatal(err)
	}
	if err := c.Sonars.Delete("abc"); err != nil {
		t.Fatalf("Delete on 204: %v", err)
	}
	want := []struct{ method, path string }{
		{"GET", "/v1/sonars"}, {"GET", "/v1/sonars/01j0ksyw6tgwfs0zmhhbc42kwa"}, {"GET", "/v1/sonars/odd%2Fid"}, {"DELETE", "/v1/sonars/abc"},
	}
	for i, w := range want {
		if (*calls)[i].method != w.method || (*calls)[i].rawPath != w.path {
			t.Errorf("call %d = %s %s, want %s %s", i, (*calls)[i].method, (*calls)[i].rawPath, w.method, w.path)
		}
	}
}

func TestSonarsUpdatePatchesOnlyGivenFieldsAndCanClearTheCap(t *testing.T) {
	c, calls := newSonarServer(t)
	if _, err := c.Sonars.Update("abc", SonarRequest{Name: "renamed"}); err != nil {
		t.Fatal(err)
	}
	if (*calls)[0].method != "PATCH" || (*calls)[0].rawPath != "/v1/sonars/abc" || string((*calls)[0].body) != `{"name":"renamed"}` {
		t.Errorf("got %s %s %s", (*calls)[0].method, (*calls)[0].rawPath, (*calls)[0].body)
	}
	// clearing needs an explicit null, which omitempty cannot express: use a map
	if _, err := c.Sonars.Update("abc", map[string]interface{}{"daily_cap": nil, "enabled": false}); err != nil {
		t.Fatal(err)
	}
	if string((*calls)[1].body) != `{"daily_cap":null,"enabled":false}` {
		t.Errorf("clear body = %s", (*calls)[1].body)
	}
}

func TestSonarsMatchesAndPreviewParams(t *testing.T) {
	c, calls := newSonarServer(t,
		`{"matches":[{"id":3,"sonar_id":"abc","post_id":"p3","matched_at":"2026-09-10T00:00:00Z","why":{"matched_by":"hashtag"},"delivered":true}],"next_before":3}`,
		`{}`,
		`{"window_days":7,"total":42,"per_day":[{"day":"2026-09-09T00:00:00.000Z","count":42}],"samples":[{"id":"x","service":"bluesky"}]}`,
	)
	raw, err := c.Sonars.Matches("abc", 0, 0)
	if err != nil {
		t.Fatal(err)
	}
	if (*calls)[0].rawPath != "/v1/sonars/abc/matches" || (*calls)[0].query != "" {
		t.Errorf("first page: %s ? %s", (*calls)[0].rawPath, (*calls)[0].query)
	}
	var page SonarMatchPage
	if err := json.Unmarshal(raw, &page); err != nil {
		t.Fatal(err)
	}
	if len(page.Matches) != 1 || page.Matches[0].PostID != "p3" || !page.Matches[0].Delivered || page.NextBefore == nil || *page.NextBefore != 3 {
		t.Errorf("page = %+v", page)
	}
	if _, err := c.Sonars.Matches("abc", *page.NextBefore, 10); err != nil {
		t.Fatal(err)
	}
	if (*calls)[1].query != "before=3&limit=10" {
		t.Errorf("second page query = %s", (*calls)[1].query)
	}

	raw, err = c.Sonars.Preview(sonarSpec, 7)
	if err != nil {
		t.Fatal(err)
	}
	if (*calls)[2].method != "POST" || (*calls)[2].rawPath != "/v1/sonars/preview" || (*calls)[2].query != "days=7" {
		t.Errorf("preview: %s %s ? %s", (*calls)[2].method, (*calls)[2].rawPath, (*calls)[2].query)
	}
	if string((*calls)[2].body) != `{"subject":{"hashtags":["#opensearch"]},"surfaces":["bluesky","mastodon"]}` {
		t.Errorf("preview body = %s", (*calls)[2].body)
	}
	var preview SonarPreview
	if err := json.Unmarshal(raw, &preview); err != nil {
		t.Fatal(err)
	}
	if preview.WindowDays != 7 || preview.Total != 42 || len(preview.PerDay) != 1 || preview.Samples[0].Service == nil || *preview.Samples[0].Service != "bluesky" {
		t.Errorf("preview = %+v", preview)
	}
	if _, err := c.Sonars.Preview(sonarSpec, 0); err != nil {
		t.Fatal(err)
	}
	if (*calls)[3].query != "" {
		t.Errorf("days <= 0 should omit the param, got %q", (*calls)[3].query)
	}
}

func TestSonarDecodesTheWireShape(t *testing.T) {
	var s Sonar
	err := json.Unmarshal([]byte(`{"id":"abc","owner_id":"o","name":"n","enabled":true,
		"spec":{"subject":{"hashtags":["#opensearch"]},"surfaces":["bluesky"]},"cadence":"instant",
		"channels":[{"type":"push"}],"daily_cap":null,"tier":"free","created":"2026-09-10T00:00:00Z","updated":"2026-09-10T00:00:00Z"}`), &s)
	if err != nil {
		t.Fatal(err)
	}
	if s.ID != "abc" || s.Spec.Subject.Hashtags[0] != "#opensearch" || s.Channels[0].Type != "push" || s.DailyCap != nil || s.Tier == nil || *s.Tier != "free" {
		t.Errorf("sonar = %+v", s)
	}
}
